android/src/main/java/com/swmansion/reanimated/layoutReanimation/AnimationsManager.java

```diff
@@ -512,16 +356,150 @@ public void updateLayout(
                 + parentTag
                 + " as a parent, but its Manager doesn't implement IViewManagerWithChildren");
       }
-      if (parentViewManagerWithChildren != null
-          && !parentViewManagerWithChildren.needsCustomLayoutForChildren()) {
+      if (!parentViewManagerWithChildren.needsCustomLayoutForChildren()) {
         viewToUpdate.layout(x, y, x + width, y + height);
       }
     } else {
       viewToUpdate.layout(x, y, x + width, y + height);
     }
   }
 
+  public boolean hasAnimationForTag(int tag, String type) {
+    return mNativeMethodsHolder.hasAnimationForTag(tag, type);
+  }
+
   public boolean isLayoutAnimationEnabled() {
     return mNativeMethodsHolder != null && mNativeMethodsHolder.isLayoutAnimationEnabled();
   }
+
+  private boolean removeOrAnimateExitRecursive(View view, ViewGroup parent, boolean shouldRemove) {
+    int tag = view.getId();
+    boolean hasExitAnimation = hasAnimationForTag(tag, "exiting");
+    boolean wantAnimateExit = hasExitAnimation;
+
+    // we might want to keep this view around
+    // because one of the (children's) children
+    // has an exiting animation
+    if (view instanceof ViewGroup) {
+      ViewGroup viewGroup = (ViewGroup) view;
+      ArrayList<View> children = new ArrayList<>();
+      for (int i = 0; i < viewGroup.getChildCount(); i++) {
+        View child = viewGroup.getChildAt(i);
+        children.add(child);
+      }
+      for (View child : children) {
+        if (removeOrAnimateExitRecursive(child, viewGroup, shouldRemove && !hasExitAnimation)) {
```

Comment:
I wonder if this is inline with the way this used to work. Not sure if a cascading remove was actually happening before. Is this implemented the same way on iOS? Maybe if all the children have no animation we could skip detaching from parent and only call drop method? They'll be removed from the hierarchy as a part of their parent's removal

Reply:
I don't think it worked like this before, but I'd say it's easier than trying to identify roots of non-animated subtrees and then removing them separately.
We can try changing this at some point in the future and seeing if it changes anything.